### PR TITLE
Improve Typographer Edge Cases

### PR DESCRIPTION
**Pull request by natemoo-re**
I mentioned I wanted to help with https://github.com/yuin/goldmark/issues/180 quite a while ago. I finally found time to do it!

This PR fixes various edge cases with the Typographer extension, with test cases borrowed from @brycewray's https://gm-typographer.vercel.app/.

Closes https://github.com/yuin/goldmark/issues/180, https://github.com/gohugoio/hugo/issues/9397.

cc @brycewray @skyfaller

### CODE REVIEW

File: extension/_test/typographer.txt [hunk 1]
```diff
@@ -64,3 +64,73 @@ reported "issue 1 (IE-only)", "issue 2", 'issue3 (FF-only)', 'issue4'
 //- - - - - - - - -//
 <p>&ldquo;Monitor 21&quot;&rdquo; and &ldquo;Monitor&rdquo;&quot;</p>
 //= = = = = = = = = = = = = = = = = = = = = = = =//
+
+9: Closing quotation marks within italics
```
Comment (natemoo-re) [inline extension/_test/typographer.txt:68]: These test cases are all borrowed from https://gm-typographer.vercel.app/!

**Comment by yuin** [comment]
@natemoo-re Thanks for your contribution. Open source project should be like this.

File: extension/_test/typographer.txt [hunk 2]
```diff
@@ -64,3 +64,73 @@ reported "issue 1 (IE-only)", "issue 2", 'issue3 (FF-only)', 'issue4'
...
+//= = = = = = = = = = = = = = = = = = = = = = = =//
+
+10: Closing quotation marks within boldfacing
+//- - - - - - - - -//
+**"At first, things were not clear."**
+//- - - - - - - - -//
+<p><strong>&ldquo;At first, things were not clear.&rdquo;</strong></p>
+//= = = = = = = = = = = = = = = = = = = = = = = =//
+
+11: Closing quotation marks within boldfacing and italics
```
Comment (jasikpark) [inline extension/_test/typographer.txt:82]: Are test cases 11&12 duplicated?
